### PR TITLE
[CALCITE-985] Validate MERGE

### PR DESCRIPTION
**Pull request by JiajunBernoulli**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 1]
```diff
@@ -2256,11 +2256,16 @@ protected void registerNamespace(
       @Nullable String alias,
       SqlValidatorNamespace ns,
       boolean forceNullable) {
-    namespaces.put(requireNonNull(ns.getNode(), () -> "ns.getNode() for " + ns), ns);
+    final SqlNode sqlNode = requireNonNull(ns.getNode(), () -> "ns.getNode() for " + ns);
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:2259]: Consider removing the `sqlNode` variable. After you've called `requireNonNull(ns.getNode())` the validator will no longer complain.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 2]
```diff
@@ -5269,6 +5274,20 @@ private static SqlNode getNthExpr(SqlNode query, int ordinal, int sourceCount) {
     SqlInsert insertCallAfterValidate = call.getInsertCall();
     if (insertCallAfterValidate != null) {
       validateInsert(insertCallAfterValidate);
+      // Validate NULL
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:5277]: I didn't understand the 'Validate NULL' comment at first. 'Throw if select list contains NULL literal and target is NOT NULL' might be better.

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 3]
```diff
@@ -3076,16 +3076,14 @@ void checkCorrelatedMapSubQuery(boolean expand) {
...
   @Test void testMerge() {
-    final String sql = "merge into emp as target\n"
-        + "using (select * from emp where deptno = 30) as source\n"
-        + "on target.empno = source.empno\n"
-        + "when matched then\n"
-        + "  update set sal = sal + source.sal\n"
-        + "when not matched then\n"
-        + "  insert (empno, deptno, sal)\n"
-        + "  values (source.empno, source.deptno, source.sal)";
+    final String sql = "merge into empnullables e "
```
Comment (julianhyde) [inline core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:3080]: Can you end each line of SQL with "\n" (and no space). It makes the SQL easier to read inside `SqlToRelConverterTest.xml`.

**Review by julianhyde — CHANGES_REQUESTED** [review]
I requested some cosmetic changes. With those, +1 to merge without further review.

Thank you for doing this -- the case has been open for many years!

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 2 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Good idea.

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java (hunk 3 above)
Comment (JiajunBernoulli) [reply core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java]: Done

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 1 above)
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Removed `sqlNode`.

**Comment by JiajunBernoulli** [comment]
Closed #3283

**Pull request merged**